File: 0_4.programmers/kakao#2/java/KAKAO16.java
import java.util.*;
public class KAKAO16 {
    class Solution {
        public int solution(int cacheSize, String[] cities) {
            int answer = 0;
            LinkedList<String> q = new LinkedList<>();
            
            for (String city : cities) {
                city = city.toLowerCase();
                if (q.contains(city)) {
                    q.remove(city);
                    q.addLast(city);
                    answer++;
                } else {
                    q.addLast(city);
                    
                    if (q.size() > cacheSize) {
                        q.removeFirst();
                    }
                    answer += 5;
                }
            }
            return answer;
        }
    }
}
/**
 * [2018 카카오 공채] 캐시
 * 해결: O
 * 
 */

File: 0_4.programmers/kakao#2/python/KAKAO16.py
from collections import deque
def solution(cacheSize, cities):
    answer = 0
    cache = deque()
    
    for city in cities:
        city = city.lower()
        # cache miss
        if city not in cache:
            cache.append(city)
            
            if len(cache) > cacheSize:
                cache.popleft()
            
            answer += 5
        # cache hit
        else:
            cache.remove(city)
            cache.append(city)
            answer += 1
            
    return answer
"""
[2018 카카오 공채] 캐시
해결: O
"""